drivers: video: mt9m114: Add vertical and horizontal flip control

jeronimoagullo opened the pull request:
This pull request gives support to flip the camera mt9m114 images horizontally and vertically. For this aim, it allows to modify camera buffer reading mode. 

This PR provides support for the use of `video_set_ctrl` function with the Camera class control IDs of `VIDEO_CID_HFLIP` and `VIDEO_CID_VFLIP`.

It was proposed by @ngphibang during a conversation. thanks!

ngphibang commented:
@jeronimoagullo : Thank you for the commit !
Could you point out the platform you tested this control and some hflipped / vflipped images of it ?

jeronimoagullo commented:
yes, @ngphibang I used the NXP imxRT1064 evaluation board. Besides, to check it out easily I modified the `samples/drivers/video/tcpserversink` sample to test it.

josuah commented on drivers/video/mt9m114.c:
```diff
@@ -461,12 +503,35 @@ static int mt9m114_get_caps(const struct device *dev, enum video_endpoint_id ep,
+{
+	int ret = 0;
+
+	switch (cid) {
+	case VIDEO_CID_HFLIP:
+		ret |= mt9m114_set_horizontal_mirror(dev, (int)value);
+		break;
+	case VIDEO_CID_VFLIP:
+		ret |= mt9m114_set_vertical_flip(dev, (int)value);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	/* Apply Config */
+	ret |= mt9m114_set_state(dev, MT9M114_SYS_STATE_ENTER_CONFIG_CHANGE);
```
It is true that the `ret |= ...` is everywhere in sensor drivers, but it might introduce small bugs in error handling. I.e. `-EIO | -EAGAIN == -EPERM` (`-5 | -11 == -1`).

It would be almost as concise to do it that way:

```suggestion
	switch (cid) {
	case VIDEO_CID_HFLIP:
		ret = mt9m114_set_horizontal_mirror(dev, (int)value);
		break;
	case VIDEO_CID_VFLIP:
		ret = mt9m114_set_vertical_flip(dev, (int)value);
		break;
	default:
		return -ENOTSUP;
	}
	if (ret < 0) {
		return ret;
	}

	/* Apply Config */
	return mt9m114_set_state(dev, MT9M114_SYS_STATE_ENTER_CONFIG_CHANGE);
```

Is it reasonable enough to switch?

An alternative is to convert all sensor drivers at once in a big refactoring when an opportunity shows-up (i.e. new APIs requiring modification anyway)...

josuah commented on drivers/video/mt9m114.c:
```diff
@@ -381,6 +383,46 @@ static int mt9m114_set_output_format(const struct device *dev, int pixel_format)
+		reg &= ~BIT(0);
+	}
+
+	ret |= mt9m114_write_reg(dev, MT9M114_CAM_SENSOR_CTRL_READ_MODE, sizeof(reg),
+				&reg);
```
Same proposal as below to turn this into `ret = ` and immediately checking the error, which is a bit more verbose but helps avoiding subtle bugs in applications (i.e. checking if some operation is not supported (no point in retrying), or just some failure on I/O (a ladybug accidentally shorted the `I2C.SDA` and `GND` lines, ladybug forgiven and new attempt issued).

josuah reviewed (changes requested):
Thank you! Having the image in correct orientation is always needed...
I proposed some minor modification about error handling, but that is really not critical.

jeronimoagullo commented:
@josuah You are totally right! I was following the code style of the video drivers and I did't figure out the possibility of confusion due to error mess. I have modified it and push, please check it out, thanks!

josuah reviewed (dismissed):
Thank you for the change! It looks ready to land to me.
I was also using `ret |=` before Zephyr, as when the only error code is `-1` this works.

kartben commented on drivers/video/mt9m114.c:
```diff
@@ -461,12 +503,36 @@ static int mt9m114_get_caps(const struct device *dev, enum video_endpoint_id ep,
+
+	switch (cid) {
+	case VIDEO_CID_HFLIP:
+		ret = mt9m114_set_horizontal_mirror(dev, (int)value);
```
s/mirror/flip/

jeronimoagullo replied:
I really like the `vim` format of your comment!

kartben replied:
:)
Note that it of course doesn't apply if you end up refactoring the two functions into one

ngphibang replied:
Hi @kartben , as you said, this doesn't apply anymore due to refactoring (implemented differently), could you resolve the Change Request to help moving this PR forwards ?

kartben commented on drivers/video/mt9m114.c:
```diff
@@ -381,6 +383,46 @@ static int mt9m114_set_output_format(const struct device *dev, int pixel_format)
+	}
+
+	if (enable) {
+		reg |= BIT(0);
```
Avoid magic numbers and #define this properly

kartben commented on drivers/video/mt9m114.c:
```diff
@@ -381,6 +383,46 @@ static int mt9m114_set_output_format(const struct device *dev, int pixel_format)
+	}
+
+	if (enable) {
+		reg |= BIT(1);
```
Ditto

kartben commented on drivers/video/mt9m114.c:
```diff
@@ -381,6 +383,46 @@ static int mt9m114_set_output_format(const struct device *dev, int pixel_format)
+	return mt9m114_write_reg(dev, MT9M114_CAM_SENSOR_CTRL_READ_MODE, sizeof(reg), &reg);
+}
+
+static int mt9m114_set_vertical_flip(const struct device *dev, int enable)
```
Unless I missed something the two functions are after all essentially duplicated? Maybe make it a common set_orientation function?

jeronimoagullo replied:
You are right, I used two different functions like other camera driver does, but in this case, the functions may be the same adding just the bit that you want to change. Thus, I will merge them into just one function

ngphibang commented:
The VIDEO_CID_HFLIP/VIDEO_CID_VFLIP CID naming are not changed in https://github.com/zephyrproject-rtos/zephyr/pull/78802 so I think it will not have conflicts.

BTW, I needs to test this PR on RT1064 before giving some reviews.

jeronimoagullo commented:
> Note that this will have conflicts when #78802 gets merged

Are you sure? I have just added flip control. So, I use only from `video-contro.h` the `VIDEO_CID_HFLIP` and `VIDEO_CID_VFLIP` macros which are not modified in #78802. Thus, for the moment, both PRs shouldn't have conflicts, right?

kartben commented:
> > Note that this will have conflicts when #78802 gets merged
> 
> Are you sure? I have just added flip control. So, I use only from `video-contro.h` the `VIDEO_CID_HFLIP` and `VIDEO_CID_VFLIP` macros which are not modified in #78802. Thus, for the moment, both PRs shouldn't have conflicts, right?

ah yes sorry :) I thought all the macros had been renamed, but turns out these ones didn't have a category prefix in the first place. Sorry!

ngphibang commented on drivers/video/mt9m114.c:
```diff
@@ -53,6 +55,10 @@ LOG_MODULE_REGISTER(video_mt9m114, CONFIG_VIDEO_LOG_LEVEL);
 #define MT9M114_CAM_OUTPUT_FORMAT_FORMAT_RGB (1 << 8)
 
+/* Camera control masks */
+#define MT9M114_CAM_CONTROL_HORIZ_FLIP_EN    BIT(0)
```
I suggest : `MT9M114_CAM_SENSOR_CTRL_HORZ_FLIP_EN` to be coherent with `MT9M114_CAM_SENSOR_CTRL_READ_MODE` and `MT9M114_CAM_SENSOR_CTRL_VERT_FLIP_EN`

ngphibang commented on drivers/video/mt9m114.c:
```diff
@@ -53,6 +55,10 @@ LOG_MODULE_REGISTER(video_mt9m114, CONFIG_VIDEO_LOG_LEVEL);
 
+/* Camera control masks */
+#define MT9M114_CAM_CONTROL_HORIZ_FLIP_EN    BIT(0)
+#define MT9M114_CAM_CONTROL_VERT_FLIP_EN     BIT(1)
```
I suggest : `MT9M114_CAM_SENSOR_CTRL_VERT_FLIP_EN` to be coherent with `MT9M114_CAM_SENSOR_CTRL_READ_MODE`

ngphibang commented on drivers/video/mt9m114.c:
```diff
@@ -381,6 +387,26 @@ static int mt9m114_set_output_format(const struct device *dev, int pixel_format)
 	return ret;
 }
 
+static int mt9m114_set_orientation(const struct device *dev, int enable, uint8_t mask)
```
I would drop this function. Just use `mt9m114_modify_reg()` directly as suggested in comments below

josuah replied:
Now that `mt9m114_modify_reg()` is used everywhere, this function can be removed.
Thanks for the modifications!

ngphibang commented on drivers/video/mt9m114.c:
```diff
@@ -461,12 +487,36 @@ static int mt9m114_get_caps(const struct device *dev, enum video_endpoint_id ep,
+
+	switch (cid) {
+	case VIDEO_CID_HFLIP:
+		ret = mt9m114_set_orientation(dev, (int)value, MT9M114_CAM_CONTROL_HORIZ_FLIP_EN);
```
`mt9m114_modify_reg(dev, MT9M114_CAM_SENSOR_CTRL_READ_MODE, MT9M114_CAM_SENSOR_CTRL_HORZ_FLIP_EN, (int)value ? MT9M114_CAM_SENSOR_CTRL_HORZ_FLIP_EN, 0);`

jeronimoagullo replied:
I see your point, but it is not feasible right now due to the definition of `mt9m114_modify_reg` function which works only with 8 bit registers. In this case, the `MT9M114_CAM_SENSOR_CTRL_READ_MODE` register is 16 bit.

I see two alternatives:
1. rewrite function to work with 8, 16 and 32 bit registers
2. write a function for each register size

Then, it would be feasible

ngphibang replied:
Ah, sorry, I just doublechecked. Indeed, inside `mt9m114_modify_reg` , it calls read/write reg on 8 bits only

	uint8_t newVal;
	int ret = mt9m114_read_reg(dev, addr, sizeof(oldVal), &oldVal);

ngphibang replied:
So if possible, could you modify also `mt9m114_modify_reg` so that it supports both 8 bits and 16 bits ? then `mt9m114_modify_reg` is only used here and on MT9M114_RST_AND_MISC_CONTROL in software reset so the changes will not be much ?

ngphibang replied:
But it's not mandatory ... If you feel uncomfortatble, it's ok for me to let it as is for now. I will make the change later

jeronimoagullo replied:
No problem! Actually, I was thinking about it, I will do it :)

jeronimoagullo replied:
@ngphibang I have pushed a new version with a common `mt9m114_modify_reg` function:

```c
static int mt9m114_modify_reg(const struct device *dev, const uint16_t addr, 
			      uint8_t reg_size, const uint32_t mask, const uint32_t val)
{
	uint32_t oldVal = 0;
	uint32_t newVal = 0;

	int ret = mt9m114_read_reg(dev, addr, reg_size, &oldVal);

	if (ret) {
		return ret;
	}

	newVal = (oldVal & ~mask) | (val & mask);

	return mt9m114_write_reg(dev, addr, reg_size, &newVal);
}
```

The idea is to cast the value to the larger likely value (4 bytes). Then, the value is correctly read and written since they are pass as void pointer and interpreated correctly in both the read and write functions. I have check it out and works correctly, what do you think?

josuah replied:
Maybe it becomes interesting to have a [CCI (Camera Common Interface) library](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/drivers/media/v4l2-core/v4l2-cci.c) that has read/write/modify built-in for various sizes.

jeronimoagullo replied:
I think that it is beyond the scope of this pull request, but it is something interested. However, right now we are using only an i2c bus. Thus, it would be interesting in case that we can configure cameras with other buses like SPI, otherwise, I don't see the point.

josuah replied:
Yes absolutely, not related to this PR! I am thinking in the background: someone with a camera sensor knows the I2C address to toggle a feature and wants it upstream: how to make it just a 1-line diff to facilitate contribution.

I agree though: good not to introduce unneeded abstraction.

ngphibang replied:
@josuah I think adding CCI would be helpful. But even with CCI, in Linux, camera drivers still uses different i2c APIs to read / write registers, so it will not make it a "standard way" for every drivers

josuah replied:
FYI, this is now proposed here to save effort while writing new drivers:
- https://github.com/zephyrproject-rtos/zephyr/pull/87935

ngphibang commented on drivers/video/mt9m114.c:
```diff
@@ -461,12 +487,36 @@ static int mt9m114_get_caps(const struct device *dev, enum video_endpoint_id ep,
+		ret = mt9m114_set_orientation(dev, (int)value, MT9M114_CAM_CONTROL_HORIZ_FLIP_EN);
+		break;
+	case VIDEO_CID_VFLIP:
+		ret = mt9m114_set_orientation(dev, (int)value, MT9M114_CAM_CONTROL_VERT_FLIP_EN);
```
`mt9m114_modify_reg(dev, MT9M114_CAM_SENSOR_CTRL_READ_MODE, MT9M114_CAM_SENSOR_CTRL_VERT_FLIP_EN, (int)value ? MT9M114_CAM_SENSOR_CTRL_VERT_FLIP_EN, 0);`

josuah reviewed (changes requested):
One more minor step...

This fixes `mt9m114_modify_reg()`  which did not accept a `size` parameter.

jeronimoagullo commented:
@josuah check it out now, I have removed the `mt9m114_set_orientation` function and did a `git rebase` to be inline with main branch.

ngphibang reviewed (commented):
@jeronimoagullo Just a few things left then it will be OK. Thanks for the changes !

jeronimoagullo commented on drivers/video/mt9m114.c:
```diff
@@ -53,6 +55,10 @@ LOG_MODULE_REGISTER(video_mt9m114, CONFIG_VIDEO_LOG_LEVEL);
 #define MT9M114_CAM_OUTPUT_FORMAT_FORMAT_YUV (0 << 8)
 #define MT9M114_CAM_OUTPUT_FORMAT_FORMAT_RGB (1 << 8)
 
+/* Camera control masks */
```
The reason why I added this comment is to add there more camera control masks in the future (brightness, contrast, etc.)

jeronimoagullo replied:
Or just do two separated commits in this single PR

ngphibang replied:
OK, it's fine to let the comment.

jeronimoagullo commented on drivers/video/mt9m114.c:
```diff
@@ -261,20 +267,21 @@ static int mt9m114_read_reg(const struct device *dev, uint16_t reg_addr, uint8_t
 
-static int mt9m114_modify_reg(const struct device *dev, const uint16_t addr, const uint8_t mask,
-			      const uint8_t val)
+static int mt9m114_modify_reg(const struct device *dev, const uint16_t addr,
```
In that case, I should create a separate commit for it, merge it and then carry on with this PR. Do you agree?

ngphibang replied:
Yes, just split the current commit into 2 commits then force-push in this same PR.

jeronimoagullo commented:
Hi! @kartben and @loicpoulain do you agree with the final version of the Pull Request? Thanks!

loicpoulain reviewed (dismissed):
LGTM

ngphibang commented:
@jeronimoagullo Please rebase your branch on the lastest main to resolve the conflict.

ngphibang commented:
@jeronimoagullo hey, can you do a (simple) `git rebase` to resolve the conflict ? It's because the https://github.com/zephyrproject-rtos/zephyr/pull/82301 was merged.

jeronimoagullo committed:
drivers: video: mt9m114: update mt9m114_modify_reg to any register length

Update mt9m114_modify_reg to support 8, 16 and 32 bits registers

Signed-off-by: Jeronimo Agullo <jeronimoagullo97@gmail.com>

jeronimoagullo committed:
drivers: video: mt9m114: Add vertical and horizontal flip control

Add set_ctrl function API for vertical and horizontal flip control
modifying the camera read mode

Signed-off-by: Jeronimo Agullo <jeronimoagullo97@gmail.com>

ngphibang commented:
@kartben Could you revisit this ? thanks !

kartben commented:
> @kartben Could you revisit this ? thanks !

Of course, sorry!
